Download adapters and components

CremaLuca opened the pull request:
## Premise
The download phase was very much specific for each provider, leading to difficulty in spotting bugs, hard to test code, and multiple failure points.

## Features

**feat(Adapter)**: Class that holds components, performs the prepare, retrieve and atomize phases calling each one of its components per phase.
**feat(AdapterComponent)**: Implements one or more of the three phases of the download, just like a middleware. They should be as generic as possible, making them harder to test but suitable for more providers.

#### Components

**Parameter validation**: Every downloader needs parameter validation because the provider only supports some values (eg. interval "1min" or "1m"), so it's a component that given the parameter name, and a validation method applies the method to check for validity of the parameter's value.

**Request**: Performs bare HTTP requests, taking values from parameters or constant ones given in the constructor. It can build a URL by using `BASE_URL + url parameters + '?' + query parameters` (not done manually, but by the requests module). Also supports HTTP headers, request timeout time, automatic JSON of the response, request limiter.

**"Chunker"**: Splits an array of elements in an array of chunks of a given maximum size. eg `max_size=3` then `[A, B, C, D, E] - > [[A, B, C], [D, E]]`. Used when the provider supports multiple tickers but sets a maximum number of them per request.

### Notes

I also wanted to keep downloaders and the filtering/streaming structures separated, in order for the downloader part to be an MVP.

## Picture of something weird

![Oke](https://i.kym-cdn.com/photos/images/newsfeed/002/196/505/926.jpg)

CremaLuca committed:
feat(adapters): Created download adapters and components abstract classes
feat(ticker splitter): component that splits tickers in groups
test(ticker splitter)

CremaLuca committed:
feat(Parameter Validator): component that checks if a parameter is in a list of possible parameter values

CremaLuca committed:
doc(match param validation): added description
fix(Parameter validator): items() instead of enumerate()
fix(Parameter validator): not more checking parameter is defined by default
fix(match param gen): now returns the generated method
test(Parameter validator)

CremaLuca committed:
rename(parameter validator): ParamValidatorComp

CremaLuca committed:
feat(request component): performs HTTP requests with parameters given by adapter's parameters
style(adapters): formatted file

CremaLuca committed:
feat(RequestComp): added response checks and kwargs update and return

CremaLuca committed:
feat(TickerGroupHandler): component that calls preparation and retrieve for all sub-components
feat(Tradier): started adapter translation

CremaLuca committed:
feat(tradier timeseries): adapter for tradier timeseries, working
feat(TickerSplitter): parametrised output ticker groups name
feat(Request): added debug mode
fix(Request): fixed status code, fixed json response
test(tradier_timeseries)

CremaLuca committed:
fix(TickerExtractor): moved it into main module
feat(TickerExtractor): added type and length checks
test(TickerExtractor)

CremaLuca committed:
feat(tradier): alla parameters checked
feat(Pram valid.): added new datetime validation
test(param valid): added dt valid. tests

CremaLuca committed:
rename(SyncAdapter): merged with AsyncAdapter

CremaLuca committed:
doc: fixed some exception texts

CremaLuca committed:
fix(Adapter): sync true to True
del(TradierTimeseries)

CremaLuca committed:
feat(MappingComp): maps values of parameters in other values, used to reuse the same adapter for different downloads

CremaLuca committed:
Added two tests for adapters and components

CremaLuca commented:
Ayo bro let me review that for you

CremaLuca commented:
Ayo thanks bro

CremaLuca commented on otri/downloader/__init__.py:
```diff
@@ -808,3 +814,420 @@ def _post_process(self, atom: Mapping, **kwargs) -> Mapping:
+
+    def atomize(self, data_stream: ReadableStream, output_stream: WritableStream, **kwargs):
+        '''
+        Third of the download functionalities.
```
A little more descriptive?

CremaLuca replied:
Updated it 👌

CremaLuca commented on otri/downloader/__init__.py:
```diff
@@ -808,3 +814,420 @@ def _post_process(self, atom: Mapping, **kwargs) -> Mapping:
+
+        Parameters:
+            components : list[AdapterComponent]
+                Ordered list of adapter components that will be called on download. Default empty list.
```
Specify that the components will be called in the given order.

CremaLuca replied:
Also, why not using variables like django does insead of passing parameters to constructor? The parameters are static for every class.

CremaLuca replied:
I did indeed end up having components as class attributes instead of constructor, it looks incredibly cleaner.
It cannot be applied to adapters too as it's more comfortable passing parameters in the constructor rather then writing the whole class just to set the parameters.

CremaLuca commented on otri/downloader/tradier.py:
```diff
@@ -326,3 +265,79 @@ def _http_request(self, tickers: Sequence[str], timeout: float) -> Union[request
+        Parameters:
+            user_key : str
+                User request authorization key.
+            debug : bool
```
Nah man use global logger with global log-level settings, avoid this kind of things.

CremaLuca replied:
Gotchu homie 👌

CremaLuca committed:
fix: test method already defined
fix: missing exception string format parameter

CremaLuca committed:
refactor: download adapter constructor parameters as class attributes

CremaLuca committed:
refactor: A.get(x, default) is default to default in A

CremaLuca committed:
refactor: removed "debug" parameter

CremaLuca committed:
refactor: removed default parameter value out_stream

CremaLuca commented:
Still hard to read and decipher, needs some work.

CremaLuca committed:
doc: adding documentation where needed

Commit:
test: adapted tradier test to changes
refactor: applied autopeep8 formatting

Commit:
refactor: optional component's kwarg return

Commit:
fix: optional kwargs return for TicketGroupHandler too
refactor: Adapter init initialises components list if not defined

Commit:
refactor: SubAdapter instead of TickerGroupHandler and TickerExtractor

Commit:
refactor: optional url in request component

Commit:
refactor: .format to f-strings

Commit:
feat: request limiter in request component

ricdezen commented on otri/downloader/__init__.py:
```diff
@@ -808,3 +817,413 @@ def _post_process(self, atom: Mapping, **kwargs) -> Mapping:
+        if not isinstance(kwargs[self._tickers_name], Iterable):
+            raise ValueError("'{self._tickers_name}' parameter is not iterable, it's {type(kwargs[self._tickers_name])}")
+
+        kwargs[self._out_name] = [kwargs[self._tickers_name][i:i + self._max_count]
```
I came back to this twice before understanding that it is splitting a list (of streams??). There has to be something in `itertools` or somewhere else to do this in a clearer way.

CremaLuca replied:
What it's doing is splitting a list in chunks of size n (except for the last one) so just like
`x = [l[i:i + n] for i in range(0, len(l), n)] `
The names are indeed pretty long and it's hard to read so I should rename something.

ricdezen replied:
I remembered correctly saying there are methods to slice a list in (almost) equal parts, but couldn't find one for slicing into fixed length parts. By the way, are you sure this raises no error? If I have a max count of 5, and 13 tickers, wouldn't this crash when taking the third slice?

CremaLuca replied:
Nope it shouldn't raise any error, when slicing if the second operator is bigger than the length of the list it will automatically take it up to the list length

```py
l=[1,2,3,4,5]
print(l[3:200]) # prints [4,5]
```

ricdezen replied:
til

ricdezen commented on otri/downloader/__init__.py:
```diff
@@ -808,3 +817,413 @@ def _post_process(self, atom: Mapping, **kwargs) -> Mapping:
+    Can be used to reuse the same adapter for different downloads
+    '''
+
+    def __init__(self, key: str, value_mapping: Mapping[str, Set[str]], required: bool = True):
```
I may or may not be tripping right now but I think this `value_mapping` could make more sense in reverse, as in mapping values to their translation, so that you can avoid 99% of the work inside `prepare` below.

Something like:
```python
elif self._key in kwargs:
    if kwargs[self._key] not in self._value_mapping:
        raise ValueError(f"Parameter's '{self._key,}' value {kwargs[self._key]} is not in {self._value_mapping}")
    kwargs[self._key] = self._value_mapping[self._key]
```

Considering you take the "first" occurrence in the mapping, which is not known in advance since mappings are not sorted by key, I think having more than one occurrence of the values to translate is not needed and should actually be avoided, hence why having the values to translate as keys makes sense in my head.

EDIT: typo

Commit:
test: tested subAdapter

ricdezen reviewed (commented):
Seems sort of fine overall, I only have two doubts I pointed out. Also I believe docstrings could be a bit more generous with explanations/examples. I don't think I would have understood what some of the components do if I could not have looked directly into the code.

Commit:
refactor: renamed confusing names in ticker splitting method

Commit:
feat: added param transforms to request component
feat: changed tradier metadata to adapters

Commit:
refactor: imports cleanup

Commit:
refactor: yahoo timeseries as adapter
fix: time_handler handles non-existing seconds
feat: Renaming component

Commit:
refactor: ParamValidation static methods do not require key as method first parameter anymore

Commit:
feat: range as yf timeseries parameter
fix: date validation properly uses required
fix: datetimetoepoch component won't translate datetime if it's None

Commit:
refactor: removed useless ticker grouper where only one ticker is supported

Commit:
doc: updated documentation and renamed TickerGrouper to TickerChunk

Commit:
feat: tradier adder provider and ticker to atoms
test: tradier checking provider and ticker in atoms
test: fixed some adapter component tests

Commit:
refactor: removing stream dependency from downloaders
fix: Tradier checking header presence before using it

Commit:
wip: Adapter component only has one method, Adapter takes two component lists
delete: unused components, for clarity
rename: TickerChunkComp to ChunkerComp (more generic)
move: Parameter validation methods factory to validators file

Commit:
reafctor: updated SubAdapter and RequestComp

Commit:
wip: updating tradier, yf and tests

Commit:
refactor: functioning tradier and yf adapters

Commit:
refactor: removed add_component methods

Commit:
refactor: download independent from stream
refactor: SubAdapter does not make a difference between prep and ret components

Commit:
fix: atomizers passing multiple times over data
test: updated components tests

Commit:
refactor: migrated yahoo metadata to adapters

CremaLuca commented:
Something that will need some work is how to pass a limiter to the request component at initialization, as of now the limiter is instantiated when the file is parsed and cannot be changed.

Commit:
cleanup: removed MetadataDownloader class

Commit:
refactor: RealtimeAdapter for continuos requests
cleanup: removed RealtimeDownloader and TradierRealtime

Commit:
fix: added provider to all adapters output

CremaLuca commented on otri/downloader/__init__.py:
```diff
@@ -546,265 +551,328 @@ def _set_chain_max_attempts(self, max_attempts: int):
-        postprocessed_atom = self._post_process(atom=prepared_atom, ticker=ticker)
+class SubAdapter(AdapterComponent, Adapter):
+    '''
+    Performs the preparation and retrieval for every element in a parameter value (if an it's an iterable).
```
```suggestion
    Performs the preparation and retrieval for each element in an iterable given its parameter name.
```

CremaLuca commented on otri/downloader/realtime.py:
```diff
@@ -0,0 +1,45 @@
+class RealtimeAdapter(Adapter):
+    '''
+    Adapter to continuosly download data from a provider. Runs preparation component only once.
+    Handles thread signals to be stopped (sigterm, sighup and siging) or by calling stop().
```
```suggestion
    Handles thread signals to be stopped (sigterm, sighup and sigint) or by calling stop().
```

Commit:
feat: yahoo options to adapter

CremaLuca committed:
Update otri/downloader/__init__.py

CremaLuca committed:
Update otri/downloader/realtime.py
